homeassistant/components/roomba/sensor.py

```diff
@@ -63,3 +81,194 @@ def icon(self):
     def native_value(self):
         """Return the state of the sensor."""
         return self._battery_level
+
+
+class CleaningTime(IRobotEntity, SensorEntity):
+    """Class to hold Roomba Sensor basic info."""
+
+    _attr_entity_category = EntityCategory.DIAGNOSTIC
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return f"{self._name} total cleaning time"
+
+    @property
+    def unique_id(self):
+        """Return the ID of this sensor."""
+        return f"total_cleaning_time_{self._blid}"
+
+    @property
+    def native_unit_of_measurement(self):
+        """Return the unit_of_measurement of the device."""
+        return "hours"
+
+    @property
+    def icon(self):
+        """Return the icon for the cleaning time."""
+        return "mdi:clock"
+
+    @property
+    def native_value(self):
+        """Return the state of the sensor."""
+        hours = self._run_stats.get("hr")
+        # minutes = self._cleaning_time.get("min")
+
+        # minutes_total = minutes + (hours * 60)
+
+        return hours
+
+
+class TotalMissions(IRobotEntity, SensorEntity):
```

Comment:
Most of these Mission classes are copypasta. Could you factor out into a generic one and pass in the few differences in the constructor?

Reply:
How exactly should I do that?
I tried it like so:
```python
    roomba_total_missions = Missions(roomba, blid, "total", "nMssn")
    roomba_success_missions = Missions(roomba, blid, "successful", "nMssnOk")
    roomba_canceled_missions = Missions(roomba, blid, "canceled", "nMssnC")
    roomba_failed_missions = Missions(roomba, blid, "failed", "nMssnF")
```

```python
class Missions(IRobotEntity, SensorEntity, MissionType, MissionValueString):
    """Class to hold the Roomba missions info."""

    _attr_entity_category = EntityCategory.DIAGNOSTIC

    @property
    def name(self):
        """Return the name of the sensor."""
        return f"{self._name} missions {MissionType}"

    @property
    def unique_id(self):
        """Return the ID of this sensor."""
        return f"total_missions_{self._blid}"

    @property
    def icon(self):
        """Return the counter icon."""
        return "mdi:counter"

    @property
    def native_unit_of_measurement(self):
        """Return the unit_of_measurement of the device."""
        return ""

    @property
    def native_value(self):
        """Return the state of the sensor."""
        return self._mission_stats.get(MissionValueString)
```

but that gives me an error in the IDE and I also don't think this correct.

Reply:
The constructor takes the form of the magic function `__init__`, so leave the class inheritance as it was with `IRobotEntity, SensorEntity`, and `def __init__(self, mission_type, mission_value_string):` You may also want to call the `super` function.

Reply:
Thanks for helping but I still don't know how to proceed.
I am generally not that bad in python but this is a bit too much at once.
I have looked up the python super functions but that didn't help me either in this case.

It would help me a lot if you can give me an example.

This is my current state:

```python
class MissionSensor(IRobotEntity, SensorEntity):
    """Class to hold the Roomba missions info."""

    def __init__(self, mission_type, mission_value_string):
        self.mission_type = mission_type
        self.mission_value_string = mission_value_string

    _attr_entity_category = EntityCategory.DIAGNOSTIC

    @property
    def name(self):
        """Return the name of the sensor."""
        return f"{self._name} missions {self.mission_type}"

    @property
    def unique_id(self):
        """Return the ID of this sensor."""
        return f"total_missions_{self._blid}"

    @property
    def icon(self):
        """Return the counter icon."""
        return "mdi:counter"

    @property
    def native_unit_of_measurement(self):
        """Return the unit_of_measurement of the device."""
        return ""

    @property
    def native_value(self):
        """Return the state of the sensor."""
        return self._mission_stats.get(self.mission_value_string)
```

My IDE says the following: __init__ method from base class 'IRobotEntity' is not called
I don't know how I can call this method.
This does not work:
```python
roomba_total_missions = MissionSensor("total", "nMssn")
```

Reply:
Check https://github.com/Xitee1/ha-core/pull/1